### PR TITLE
Improve types.

### PR DESCRIPTION
**Pull request by mshima**
- add types helpers with generics.
- implement first sourceType type with `addEntitiesToClient` api.
- implement initial Entity type and rework some properties.
- rework a couple of angular needles.
- convert angular generator to mts.

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/base-application/tasks.d.mts [hunk 1]
```diff
@@ -70,43 +55,45 @@ type ClientSource<ExtendsSelf extends ClientSource = ClientSource> = {
...
+    | 'preparingEachEntityTaskParam'
+    | 'preparingEachEntityFieldTaskParam'
+    | 'preparingEachEntityRelationshipTaskParam'
+    | 'postPreparingEachEntityTaskParam'
+    | 'writingEntitiesTaskParam'
+    | 'postWritingEntitiesTaskParam',
+    ControlTaskParam & ApplicationTaskParam<Definition>
+  > & {
+    // Add additional types to each priority
+    applicationType: Definition['applicationType'];
```
Comment (Tcharl) [inline generators/base-application/tasks.d.mts:90]: I may have misundertood the feature: Isn't it confusing to mix tasks param definition and type?

File: generators/base-application/types/entity.d.mts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  fields?: Field[];
+  relationships?: Relationship[];
+
+  readOnly?: boolean;
+  embedded?: boolean;
+  skipClient?: boolean;
+  skipServer?: boolean;
+};
+
+type Entity = Required<BaseEntity> & {
```
Comment (Tcharl) [inline generators/base-application/types/entity.d.mts:35]: Awesome

File: generators/base-application/types/relationship.d.mts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type Relationship = {
+  relationshipName: string;
+} & Record<string, any>;
```
Comment (Tcharl) [inline generators/base-application/types/relationship.d.mts:22]: Enum on relationship kind?

File: generators/server/generator.mjs [hunk 4]
```diff
@@ -523,7 +523,7 @@ export default class JHipsterServerGenerator extends BaseApplicationGenerator {
                 field
               )} \nHibernate JPA 2 Metamodel does not work with Bean Validation 2 for LOB fields, so LOB validation is disabled`
             );
-            field.validation = false;
+            field.fieldValidate = false;
```
Comment (Tcharl) [inline generators/server/generator.mjs:526]: `fieldValidation` (to align with `fieldHasValidation`)?

**Review by Tcharl — COMMENTED** [review]
Awesome contribution!

Aligning terms looks to be a must (looks to be the main part of this PR, and some terms modification are not following the same logic).

Derived logic as a method within the type looks to be a future go-to (it's hard for a neophyt to know which generator phase should do what, and this approach will simplify this part)

File: generators/angular/generator.mts [hunk 5]
```diff
@@ -281,8 +236,23 @@ export default class AngularGenerator extends BaseApplicationGenerator {
    * @param {boolean} enableTranslation - If translations are enabled or not
    * @param {string} translationKeyMenu - i18n key for entry in the admin menu
    */
-  addElementToAdminMenu(routerName, iconName, enableTranslation, translationKeyMenu = _.camelCase(routerName)) {
-    this.needleApi.clientAngular.addElementToAdminMenu(routerName, iconName, enableTranslation, translationKeyMenu, this.jhiPrefix);
+  addElementToAdminMenu(routerName, iconName, enableTranslation, translationKeyMenu = _.camelCase(routerName), jhiPrefix?) {
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/angular/generator.mts:239]: Once we implement the needle at the source object.

File: generators/angular/generator.mts [hunk 6]
```diff
@@ -281,8 +236,23 @@ export default class AngularGenerator extends BaseApplicationGenerator {
...
+    this.needleApi.clientAngular.addElementToAdminMenu(routerName, iconName, enableTranslation, translationKeyMenu, jhiPrefix);
+  }
+
+  addEntitiesToMenu({ application, entities }: { application: ClientApplication; entities: Entity[] }) {
+    const filePath = `${application.clientSrcDir}app/layouts/navbar/navbar.component.html`;
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/angular/generator.mts:244]: My idea is to drop the NeedleApi, NeedlesBase and every derived class.

File: generators/angular/templates/src/main/webapp/app/entities/_entityFolder/_entityFile.model.ts.ejs [hunk 7]
```diff
@@ -19,7 +19,7 @@
 <%
 const enumImports = this.generateEntityClientEnumImports(fields);
 %>
-<%_ if (fieldsContainInstant || fieldsContainZonedDateTime || fieldsContainLocalDate) { _%>
+<%_ if (anyFieldIsInstant || anyFieldIsZonedDateTime || anyFieldIsLocalDate) { _%>
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/angular/templates/src/main/webapp/app/entities/_entityFolder/_entityFile.model.ts.ejs:22]: Sure I didn't go though everything.
```suggestion
<%_ if (anyFieldIsDateDerived) { _%>
```

File: generators/base-application/support/prepare-entity.mts [hunk 8]
```diff
@@ -122,7 +116,7 @@ function _derivedProperties(entityWithConfig) {
 
 export const entityDefaultConfig = {
   pagination: binaryOptions.DefaultValues[binaryOptions.Options.PAGINATION],
-  validation: false,
+  anyPropertyHasValidation: false,
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/support/prepare-entity.mts:119]: In this case it's valid for fields and relationships.
My idea is to make property for both fields and relationships.
Just like we can add a `Property` type with shared properties:
https://github.com/jhipster/generator-jhipster/blob/d4a25be226f8402730f06d2c40148d3cdfc362f4/generators/base-application/support/prepare-relationship.mjs#L194
https://github.com/jhipster/generator-jhipster/blob/d4a25be226f8402730f06d2c40148d3cdfc362f4/generators/base-application/support/prepare-field.mjs#L264

File: generators/base-application/support/prepare-entity.mts [hunk 9]
```diff
@@ -510,80 +503,57 @@ export function loadRequiredConfigIntoEntity(entity, config) {
...
-    }
-  });
+  preparePostEntityCommonDerivedPropertiesNotTyped(entity);
+}
 
-  entity.fields.forEach(field => {
-    const fieldType = field.fieldType;
+function preparePostEntityCommonDerivedPropertiesNotTyped(entity: any) {
+  const { relationships, fields } = entity;
+  const oneToOneRelationships = relationships.filter(({ relationshipType }) => relationshipType === 'one-to-one');
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/support/prepare-entity.mts:537]: In this case, fieldsContainNoOwnerOneToOne is used to disable spring data entity filtering due to compatibility.
Didn't go far to understand the code.
It seems that this filter is not used at all.

We should just drop or name it according the this behavior `disableBasicEntityFiltering`.

File: generators/base-application/support/prepare-entity.mts [hunk 10]
```diff
@@ -510,80 +503,57 @@ export function loadRequiredConfigIntoEntity(entity, config) {
...
       }
-    }
+      return relationshipsByType;
+    }, {});
 
-    if (Array.isArray(field.fieldValidateRules) && field.fieldValidateRules.length >= 1) {
-      entity.validation = true;
-    }
-  });
+  entity.differentRelationships = relationshipsByType;
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/support/prepare-entity.mts:554]: It already exists.
This a map of relationships by otherEntityName.
Didn't added to the type yet. Needs a better name.

File: generators/base-application/tasks.d.mts (hunk 1 above)
Comment (mshima) [reply generators/base-application/tasks.d.mts:90]: It should contain every generator customizable type, not only tasks.
Not surer it's used, but it's there for convenience.

File: generators/base-application/types/entity.d.mts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  entityTranslationKey: string;
+  entityTranslationKeyMenu: string;
+
+  i18nKeyPrefix: string;
+  i18nAlertHeaderPrefix: string;
+
+  entityApi: string;
+  entityPage: string;
+
+  anyFieldIsBigDecimal: boolean;
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/types/entity.d.mts:85]: About the method, this was already discussed.
If we convert to class, we will loose blueprint customizations.

File: generators/base-application/types/field.d.mts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type Field = {
+  fieldName: string;
+  fieldType: string;
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/types/field.d.mts:22]: Can be added later. The PR is already too big.

File: generators/base-application/types/relationship.d.mts (hunk 3 above)
Comment (mshima) [reply generators/base-application/types/relationship.d.mts:22]: Later.

File: generators/liquibase-changelogs/files.mjs [hunk 13]
```diff
@@ -109,7 +106,7 @@ export const fakeFiles = {
       ],
     },
     {
-      condition: generator => generator.entity.fieldsContainImageBlob === true || generator.entity.fieldsContainBlob === true,
+      condition: generator => generator.entity.anyFieldHasImageContentType === true || generator.entity.anyFieldIsBlobDerived === true,
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/liquibase-changelogs/files.mjs:109]: No need.
```suggestion
      condition: generator => generator.entity.anyFieldHasImageContentType || generator.entity.anyFieldIsBlobDerived,
```

File: generators/liquibase-changelogs/files.mjs [hunk 14]
```diff
@@ -119,7 +116,7 @@ export const fakeFiles = {
       ],
     },
     {
-      condition: generator => generator.entity.fieldsContainTextBlob === true,
+      condition: generator => generator.entity.anyFieldHasTextContentType === true,
```
Comment (mshima) [inline generators/liquibase-changelogs/files.mjs:119]:
```suggestion
      condition: generator => generator.entity.anyFieldHasTextContentType,
```

File: generators/server/generator.mjs (hunk 4 above)
Comment (mshima) [reply generators/server/generator.mjs:526]: Probably should be ok if there is no jdl support on `fieldValidation`.
In any case `field.validation` doesn't exist.

File: generators/base-application/types/entity.d.mts (hunk 11 above)
[the remark this answers is not in the document]
Comment (mshima) [reply generators/base-application/types/entity.d.mts:85]: The BaseEntity is something to discuss in any case would be better to share it with jdl.

**Comment by DanielFran** [comment]
@mshima Do not think that it is related to this PR, but actually I am not able to use main branch for testing

Using Node 18 LTS version and trying to use main branch to generate apps, I have this error:
````
>jhipster jdl test.jdl
INFO! Using bundled JHipster
ERROR! Named export 'isReservedTableName' not found. The requested module '../../../jdl/jhipster/reserved-keywords.js' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '../../../jdl/jhipster/reserved-keywords.js';
const { isReservedTableName } = pkg;

file:///C:/Users/danie/IdeaProjects/jhipster/generator-jhipster/dist/generators/server/support/prepare-entity.mjs:21
import { isReservedTableName } from '../../../jdl/jhipster/reserved-keywords.js';
         ^^^^^^^^^^^^^^^^^^^
SyntaxError: Named export 'isReservedTableName' not found. The requested module '../../../jdl/jhipster/reserved-keywords.js' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '../../../jdl/jhipster/reserved-keywords.js';
const { isReservedTableName } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)
````

**Comment by mshima** [comment]
@DanielFran my guess is that the dist folder is dirty with old files have you tried to remove it and rebuild?

**Comment by DanielFran** [comment]
@mshima thanks, will try it tomorrow

**Review by DanielFran — APPROVED** [review]

File: generators/base-application/support/prepare-entity.mts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tcharl) [reply generators/base-application/support/prepare-entity.mts:119]: So `field`, `relationship` and `property` (superset)? Perfect :-)

File: generators/base-application/support/prepare-entity.mts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Tcharl) [reply generators/base-application/support/prepare-entity.mts:537]: Hmm, disagrees a bit on this statement... behavior yes, but not usage. What if we had to transpose what you have in mind to all other `anyField...`? Won't work.

**Comment by DanielFran** [comment]
> @mshima thanks, will try it tomorrow

Confirmed, it was due to dist/

**Pull request merged**